src/main/drivers/castle_telemetry_decode.c

```diff
@@ -0,0 +1,206 @@
+    timerCCHandlerRec_t pwmEdgeCb;
+    volatile timCCR_t *timingChannelCCR;
+    volatile timCCR_t *directChannelCCR;
+    timerChannel_t* timer;
+    timCCR_t outputEnableTime;
+    timCCR_t pwmEdge;
+    timCCR_t saveCCR;
+    castleTelemetry_t telem[2];
+    uint8_t directChannelIndex;
+    uint8_t whichTelem; // telem we are writing to (0 or 1).
+    uint8_t telemIndex; // where in the telem struct are we?
+} castleInterrupt_t;
+
+static FAST_DATA_ZERO_INIT castleInterrupt_t castleState;
+
+void getCastleTelemetry(castleTelemetry_t* telem) {
+    ATOMIC_BLOCK(NVIC_PRIO_TIMER) {
+        memcpy(telem, &castleState.telem[castleState.whichTelem^1], sizeof(castleTelemetry_t));
+    }
+}
+
+// Both the HAL and the LL drivers generate a lot more code than we need; doing
+// direct register manipulation instead saves us several microseconds per
+// PWM cycle.
+#define IC_ENABLE_POLARITY_RISING TIM_CCER_CC1E
+#define IC_ENABLE_POLARITY_FALLING (TIM_CCER_CC1P | TIM_CCER_CC1E)
+#define IC_ENABLE_POLARITY_BOTHEDGES (TIM_CCER_CC1P | TIM_CCER_CC1NP | TIM_CCER_CC1E)
+#define OC_ENABLE_POLARITY_HIGH TIM_CCER_CC1E
+#define OC_ENABLE_POLARITY_LOW (TIM_CCER_CC1P | TIM_CCER_CC1E)
+
+#define DISABLE_CHANNEL(tim, channelIndex) (tim->CCER &= ~((timCCER_t)0xB << (channelIndex << 2)))
+// This ENABLE_CHANNEL_POLARITY only works if the relevant CCER bits
+// start out all clear, which they will if they were disabled with
+// DISABLE_CHANNEL
+#define ENABLE_CHANNEL_POLARITY(tim, channelIndex, polarity) (tim->CCER |= ((timCCER_t)polarity << (channelIndex << 2)))
+
+#define IC_FILTER_FDIV1 0
+// 15 more IC_FILTER values omitted.
+#define IC_PSC_DIV1 0
+#define IC_PSC_DIV2 TIM_CCMR1_IC1PSC_0
+#define IC_PSC_DIV4 TIM_CCMR1_IC1PSC_1
+#define IC_PSC_DIV8 TIM_CCMR1_IC1PSC
+#define IC_INPUT_DIRECT TIM_CCMR1_CC1S_0
+#define IC_INPUT_INDIRECT TIM_CCMR1_CC1S_1
+#define IC_INPUT_TRC TIM_CCMR1_CC1S
+
+#define OC_MODE_FROZEN 0
+#define OC_MODE_ACTIVE TIM_CCMR1_OC1M_0
+#define OC_MODE_INACTIVE TIM_CCMR1_OC1M_1
+#define OC_MODE_TOGGLE (TIM_CCMR1_OC1M_1 | TIM_CCMR1_OC1M_0)
+#define OC_MODE_FORCED_ACTIVE TIM_CCMR1_OC1M_2
+#define OC_MODE_FORCED_INACTIVE (TIM_CCMR1_OC1M_2 | TIM_CCMR1_OC1M_0)
+#define OC_MODE_PWM1 (TIM_CCMR1_OC1M_2 | TIM_CCMR1_OC1M_1)
+#define OC_MODE_PWM2 (TIM_CCMR1_OC1M_2 | TIM_CCMR1_OC1M_1 | TIM_CCMR1_OC1M_0)
+// modes with OCxM_3 set aren't supported here.
+#define OC_PRELOAD_DISABLE 0
+#define OC_PRELOAD_ENABLE TIM_CCMR1_OC1PE
+#define OC_FAST_DISABLE 0
+#define OC_FAST_ENABLE TIM_CCMR1_OC1FE
+
+// Must provide OC_MODE and OC_FAST and OC_PRELOAD
+// or IC_FILTER, IC_PSC, and IC_INPUT
+typedef uint32_t timCCMR_t;
+#define CAPTURE_CONFIGURE(tim, channelIndex, conf) do {                 \
+        timCCMR_t val = *(&tim->CCMR1 + ((channelIndex >> 1)&1));       \
+        val = (val & (~((timCCMR_t)0xFF)<< ((channelIndex & 1)<<3))) | ((timCCMR_t)(conf) << ((channelIndex & 1)<<3)); \
+        *(&tim->CCMR1 + ((channelIndex >> 1)&1)) = val; \
+    } while(0);
```

Comment:
No ; here
Please enclose all macro parameters in () in the body.

Reply:
fixed